src/rules/no_constant_condition.rs

```diff
@@ -0,0 +1,443 @@
+    assert_lint_err::<NoConstantCondition>(r#"while(x = 1);"#, 6);
+    assert_lint_err::<NoConstantCondition>(r#"while(function(){});"#, 6);
+    assert_lint_err::<NoConstantCondition>(r#"while(true);"#, 6);
+    assert_lint_err::<NoConstantCondition>(r#"while(() => {});"#, 6);
+    assert_lint_err::<NoConstantCondition>(r#"while(`foo`);"#, 6);
+    assert_lint_err::<NoConstantCondition>(r#"while(``);"#, 6);
+    assert_lint_err::<NoConstantCondition>(r#"while(`${'foo'}`);"#, 6);
+    assert_lint_err::<NoConstantCondition>(r#"while(`${'foo' + 'bar'}`);"#, 6);
+
+    // typeof conditions
+    assert_lint_err::<NoConstantCondition>(r#"if(typeof x){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(`${typeof x}`){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(`${''}${typeof x}`){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(
+      r#"if(typeof 'abc' === 'string'){}"#,
+      3,
+    );
+    assert_lint_err::<NoConstantCondition>(r#"if(a = typeof b){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(a, typeof b){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(
+      r#"if(typeof 'a' == 'string' || typeof 'b' == 'string'){}"#,
+      3,
+    );
+    assert_lint_err::<NoConstantCondition>(r#"while(typeof x){}"#, 6);
+
+    // void conditions
+    assert_lint_err::<NoConstantCondition>(r#"if(1 || void x);"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(void x);"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(y = void x);"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(x, void x);"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(void x === void y);"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(void x && a);"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(a && void x);"#, 3);
+
+    assert_lint_err::<NoConstantCondition>(r#"if(false && abc==='str'){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(true || abc==='str'){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if(abc==='str' || true){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(
+      r#"if(abc==='str' || true || def ==='str'){}"#,
+      3,
+    );
+    assert_lint_err::<NoConstantCondition>(r#"if(false || true){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(
+      r#"if(typeof abc==='str' || true){}"#,
+      3,
+    );
+
+    // string literals
+    assert_lint_err::<NoConstantCondition>(r#"if('str1' || 'str2'){}"#, 3);
+    assert_lint_err::<NoConstantCondition>(r#"if('str1' && 'str2'){}"#, 3);
+    // TODO(humancalico) make these test pass
+    // assert_lint_err::<NoConstantCondition>(r#"if(abc==='str' || 'str'){}"#, 3);
+    // assert_lint_err::<NoConstantCondition>(r#"if(a || 'str'){}"#, 3);
+
+    // assert_lint_err::<NoConstantCondition>(r#"function* foo(){while(true){} yield 'foo';}"#, );
+    // assert_lint_err::<NoConstantCondition>(r#"function* foo(){while(true){if (true) {yield 'foo';}}}"#, );
```

Comment:
What's the problem with these tests?

Reply:
These should work once [this](https://github.com/eslint/eslint/blob/f4d7b9e1a599346b2f21ff9de003b311b51411e6/lib/rules/no-constant-condition.js#L135-L146) condition is added